File: Windows/src/CodexMeter.Core/Services/UsageNormalizer.cs
using CodexMeter.Core.Domain;

namespace CodexMeter.Core.Services;

public sealed record UsageNormalizationState(
    TokenUsage? CumulativeHighWaterMark,
    DateTimeOffset? LastObservedAt,
    DataQuality Quality)
{
    public static UsageNormalizationState Empty { get; } = new(null, null, DataQuality.Exact);
}

public sealed record UsageNormalizationResult(
    TokenUsage? Delta,
    UsageNormalizationState State,
    string? Diagnostic);

public sealed class UsageNormalizer
{
    public UsageNormalizationResult Normalize(
        TokenObservation observation,
        UsageNormalizationState state)
    {
        if (state.LastObservedAt is not null
            && observation.OccurredAt < state.LastObservedAt.Value)
        {
            return new UsageNormalizationResult(
                null,
                state,
                "out-of-order token snapshot ignored");
        }

        if (observation.CumulativeUsage is not { } cumulative)
        {
            return new UsageNormalizationResult(
                null,
                state with { Quality = DataQuality.Partial },
                "missing cumulative token usage");
        }

        if (state.CumulativeHighWaterMark is not { } previous)
        {
            if (observation.LastUsage != cumulative)
            {
                return new UsageNormalizationResult(
                    null,
                    new UsageNormalizationState(cumulative, observation.OccurredAt, DataQuality.Partial),
                    "initial cumulative baseline is unresolved");
            }

            return new UsageNormalizationResult(
                cumulative == TokenUsage.Zero ? null : cumulative,
                new UsageNormalizationState(cumulative, observation.OccurredAt, state.Quality),
                null);
        }

        if (!cumulative.IsComponentWiseAtLeast(previous))
        {
            if (observation.LastUsage == cumulative)
            {
                return new UsageNormalizationResult(
                    cumulative == TokenUsage.Zero ? null : cumulative,
                    new UsageNormalizationState(cumulative, observation.OccurredAt, DataQuality.Partial),
                    "cumulative counter restarted");
            }

            return new UsageNormalizationResult(
                null,
                new UsageNormalizationState(
                    previous.ComponentWiseMaximum(cumulative),
                    observation.OccurredAt,
                    DataQuality.Partial),
                "cumulative token usage decreased or interleaved");
        }

        var delta = cumulative.SubtractFloorAtZero(previous);
        return new UsageNormalizationResult(
            delta == TokenUsage.Zero ? null : delta,
            new UsageNormalizationState(cumulative, observation.OccurredAt, state.Quality),
            null);
    }
}

File: Sources/CodexMeter/Sources/UsageNormalizer.swift
import Foundation

struct UsageNormalizationState: Equatable, Sendable {
    var cumulativeHighWaterMark: TokenUsage?
    var quality: DataQuality

    static let empty = UsageNormalizationState(cumulativeHighWaterMark: nil, quality: .exact)
}

struct UsageNormalizationResult: Equatable, Sendable {
    let delta: TokenUsage?
    let state: UsageNormalizationState
    let diagnostic: String?
}

struct UsageNormalizer: Sendable {
    func normalize(
        _ observation: CodexTokenObservation,
        metadata: SessionMetadata?,
        state: UsageNormalizationState
    ) -> UsageNormalizationResult {
        guard let cumulative = observation.cumulativeUsage else {
            return UsageNormalizationResult(
                delta: nil,
                state: UsageNormalizationState(
                    cumulativeHighWaterMark: state.cumulativeHighWaterMark,
                    quality: .partial
                ),
                diagnostic: "missing cumulative token usage"
            )
        }

        guard let previous = state.cumulativeHighWaterMark else {
            if metadata?.inheritsHistory == true {
                return UsageNormalizationResult(
                    delta: nil,
                    state: UsageNormalizationState(cumulativeHighWaterMark: cumulative, quality: .partial),
                    diagnostic: "inherited history baseline is unresolved"
                )
            }

            guard observation.lastUsage == cumulative else {
                return UsageNormalizationResult(
                    delta: nil,
                    state: UsageNormalizationState(cumulativeHighWaterMark: cumulative, quality: .partial),
                    diagnostic: "initial cumulative baseline is unresolved"
                )
            }

            return UsageNormalizationResult(
                delta: cumulative == .zero ? nil : cumulative,
                state: UsageNormalizationState(cumulativeHighWaterMark: cumulative, quality: state.quality),
                diagnostic: nil
            )
        }

        guard cumulative.isComponentWiseAtLeast(previous) else {
            if observation.lastUsage == cumulative {
                return UsageNormalizationResult(
                    delta: cumulative == .zero ? nil : cumulative,
                    state: UsageNormalizationState(cumulativeHighWaterMark: cumulative, quality: .partial),
                    diagnostic: "cumulative counter restarted"
                )
            }

            return UsageNormalizationResult(
                delta: nil,
                state: UsageNormalizationState(
                    cumulativeHighWaterMark: previous.componentWiseMaximum(with: cumulative),
                    quality: .partial
                ),
                diagnostic: "cumulative token usage decreased or interleaved"
            )
        }

        let delta = cumulative.subtractingFloorAtZero(previous)
        return UsageNormalizationResult(
            delta: delta == .zero ? nil : delta,
            state: UsageNormalizationState(cumulativeHighWaterMark: cumulative, quality: state.quality),
            diagnostic: nil
        )
    }
}
